modules/frontend/querysharding.go

```diff
@@ -0,0 +1,164 @@
+package frontend
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/go-kit/kit/log"
+	"github.com/weaveworks/common/user"
+
+	"github.com/grafana/tempo/modules/querier"
+)
+
+const (
+	querierPrefix  = "/querier"
+	queryDelimiter = "?"
+)
+
+func ShardingWare(queryShards int, logger log.Logger) Middleware {
+	return MiddlewareFunc(func(next Handler) Handler {
+		return shardQuery{
+			next:        next,
+			queryShards: queryShards,
+			logger:      logger,
+		}
+	})
+}
+
+type shardQuery struct {
+	next            Handler
+	queryShards     int
+	logger          log.Logger
+	blockBoundaries [][]byte
+}
+
+// Do implements Handler
+func (s shardQuery) Do(r *http.Request) (*http.Response, error) {
+	userID, err := user.ExtractOrgID(r.Context())
+	if err != nil {
+		return nil, err
+	}
+
+	// only need to initialise boundaries once
+	if len(s.blockBoundaries) == 0 {
+		s.blockBoundaries = createBlockShards(s.queryShards)
+	}
+
+	reqs := make([]*http.Request, s.queryShards)
+	for i := 0; i < s.queryShards; i++ {
```

Comment:
`i < len(s.blockBoundaries) - 1` ?
This would give `createBlockShards` the freedom to make whatever block boundaries it wants and the rest of the code will just work.